### PR TITLE
[Core] Add CuMemAllocator.discard() for tag-selective GPU memory release

### PR DESCRIPTION
**Pull request by AlanFokCo**
## What

Add a `discard(tags)` method to `CuMemAllocator` (and `XpuMemAllocator`) that releases GPU physical memory for the specified tags **without** copying data to CPU first.

## Why

The existing `sleep()` operates on **all** allocations — it offloads the specified tags to CPU and unmaps everything else. There is no way to selectively release just one tag (e.g. stale KV cache) while keeping other tags (e.g. model weights) mapped and usable on GPU.

This gap matters for two scenarios:
- **Multi-model switching**: after loading a new model, the old model's KV cache is stale and should be freed without the cost of copying it to CPU. The new model's weights must stay on GPU.
- **RL weight sync**: after receiving updated weights, the old KV cache computed with stale weights should be discarded efficiently.

## How

- `discard(tags)` iterates `pointer_to_data`, calls `unmap_and_release()` only for allocations matching the given tags, and sets a `discarded` flag on `AllocationData`.
- `sleep()` skips already-discarded allocations (prevents double `unmap_and_release` → `CUDA_ERROR_INVALID_VALUE`).
- `wake_up()` clears the `discarded` flag after remapping.

## Changes

| File | Change |
|------|--------|
| `vllm/device_allocator/__init__.py` | Add `discarded: bool` field to `AllocationData`, add `discard()` to `MemAllocator` protocol |
| `vllm/device_allocator/cumem.py` | Implement `CuMemAllocator.discard()`, update `sleep()` and `wake_up()` for `discarded` flag |
| `vllm/device_allocator/xpumem.py` | Implement `XpuMemAllocator.discard()`, same flag handling |

## Test

Verified on H20 (NVIDIA H20 96GB, CUDA 12.9) with the following test matrix:

| Test | Result |
|------|--------|
| `discard("kv_cache")` frees GPU memory | ✅ |
| Weights remain valid after discard | ✅ |
| `wake_up(["kv_cache"])` remaps after discard | ✅ |
| Weights survive full discard+wake cycle | ✅ |
| `sleep()` after `discard()` — no double-unmap crash | ✅ |
| `sleep()` + `wake_up()` restores weights correctly | ✅ |
| Double `discard()` is safe no-op | ✅ |
| `discard("nonexistent")` is safe no-op | ✅ |
| `discard(("kv_cache",))` tuple arg works | ✅ |

## Related

- RFC #34303 — multi-model switching use case
- PR #45398 — CuMemTagBackend (needs selective release primitive)
- Issue #41733 — 2026 Q2 RL roadmap "Runtime State Switching"

### CODE REVIEW

**Comment by AlanFokCo** [comment]
This is already guarded — `discard()` checks `not data.is_asleep` before calling `unmap_and_release()` (line 310), so a second `discard()` on the same tag is a no-op. Similarly, `sleep()` has `if data.is_asleep: continue` (line 253) to skip allocations that were already discarded.

**Comment by AlanFokCo** [comment]
@youkaichao This touches `CuMemAllocator` — would appreciate your review since you own this area. The change adds a `discard(tags)` method (~50 lines) that selectively releases GPU memory for specific tags without the full `sleep()` round-trip.

@andakai FYI — you mentioned in #44890 you'd rebase `release_kv_cache()` on top of this once it lands. Just rebased onto main, ready for review.

@aoshen02 This is the `discard()` primitive referenced in the CI test cases I pushed to your branch — the `discard → sleep → wake_up` state machine tests in `test_cumem_physical.py` exercise this method.

**Comment by matteso1** [comment]
@AlanFokCo really like this, and the cross-arch H20 numbers (32B in ~2s) are a strong case. discard(tags) fits naturally as a capability on the SleepModeBackend seam in #44074: a backend with supports_selective_offload plus a discard path. so the reload_model() you're building toward could land as a backend method rather than touching CuMemAllocator directly. happy to sync on the interface so your work composes in cleanly once #44074 is in. it's mergeable now, just waiting on the ready label and a review.

**Comment by AlanFokCo** [comment]
@matteso1 Thanks — glad it maps cleanly to the backend seam. I was thinking the same direction: once #44074 lands, `discard(tags)` becomes a method on `SleepModeBackend` with a `supports_selective_release` capability flag, so backends that only do full suspend/resume can declare they don't support it and callers fall back gracefully.

Rough sketch of how it could compose:

- `SleepModeBackend.suspend(tags)` — offload to CPU (existing)
- `SleepModeBackend.resume(tags)` — restore from CPU (existing)
- `SleepModeBackend.discard(tags)` — release without backup (new, opt-in via capability)

The `CuMemBackend` implementation would delegate to the allocator's `discard()` method directly. Backends like `cuda_checkpoint` that work at a different granularity can return `supports_selective_release = False` and the orchestration layer handles the fallback (full suspend + reload).

Happy to draft a concrete interface proposal once #44074 passes CI. Would a short doc or a GitHub discussion thread work better for iterating on the API shape?

**Comment by AlanFokCo** [comment]
Rebased onto latest main (now includes #44074 SleepModeBackend). No conflicts — the diff is unchanged (+42 lines, 2 files).

With the backend abstraction now landed, `discard(tags)` can be wired into `SleepModeBackend` as a third operation alongside `suspend`/`resume` in a follow-up PR, as discussed with @matteso1 above.

@youkaichao @aoshen02 this is ready for review — small change, fully tested on H20.

**Comment by andakai** [comment]
Hi @AlanFokCo , could you add a test for this discard tag function? Maybe you can reference to what I did here https://github.com/vllm-project/vllm/pull/44890/changes#diff-23a34cf78668b51531d6cc3435196dec1754d0430485e82e6c28900ef27c237b. I think you can also add
```
gc.collect()
torch.cuda.empty_cache()
```
in the end to be consistent with sleep in case the memory won't be released. By the way, could you also simplify the comment of this function?

**Comment by andakai** [comment]
`XpuMemAllocator`'s implementation is missed in this pr now?
Oh I just saw you pushed it.

**Comment by AlanFokCo** [comment]
Good catch — just pushed a fix. Added `discard()` to `XpuMemAllocator` with the same semantics, plus `is_asleep` state tracking in its `sleep()`/`wake_up()` paths to stay consistent with `CuMemAllocator`. Should be passing mypy now.

File: vllm/device_allocator/xpumem.py [hunk 1]
```diff
@@ -216,11 +220,38 @@ def sleep(self, offload_tags: tuple[str, ...] | str | None = None) -> None:
...
             xpu_empty_cache()
 
+    def discard(self, tags: tuple[str, ...] | str) -> None:
+        """Selectively release XPU memory for the given tags without
+        CPU backup.  Other tags remain mapped and usable."""
+        if isinstance(tags, str):
+            tags = (tags,)
+
+        discarded_bytes = 0
+        for ptr, data in self.pointer_to_data.items():
```
Comment (andakai) [inline vllm/device_allocator/xpumem.py]: same as cumem.py

File: vllm/device_allocator/cumem.py [hunk 2]
```diff
@@ -280,6 +282,30 @@ def sleep(self, offload_tags: tuple[str, ...] | str | None = None) -> None:
...
         torch.cuda.empty_cache()
 
+    def discard(self, tags: tuple[str, ...] | str) -> None:
+        """Selectively release GPU memory for the given tags without
+        CPU backup.  Other tags remain mapped and usable."""
+        if isinstance(tags, str):
+            tags = (tags,)
+
+        discarded_bytes = 0
+        for ptr, data in self.pointer_to_data.items():
```
Comment (andakai) [inline vllm/device_allocator/cumem.py]: I think use `pointer_to_data.values` is fine here. `ptr` is not used.

**Review by andakai — COMMENTED** [review]
small comments

File: tests/basic_correctness/test_mem.py [hunk 3]
```diff
@@ -75,6 +75,41 @@ def test_basic_cumem():
...
+    assert torch.allclose(weights, torch.ones_like(weights))
+
+    # Wake up and verify kv_cache is remapped (zeroed content)
+    allocator.wake_up(["kv_cache"])
+    # After wake_up the VA is remapped; content is not preserved
+    # but the allocation is valid
+    assert kv.shape == (512, 512)
+
+    # Full sleep/wake cycle still works after discard
+    allocator.sleep(offload_tags=("weights", "kv_cache"))
```
Comment (andakai) [inline tests/basic_correctness/test_mem.py]: Why kvcache in offload_tags here? If you want to test the previous sleep/wake cycle, sleep level 1 only has offload_tags=weight. sleep level 2 has offload_tags=()

File: vllm/device_allocator/xpumem.py (hunk 1 above)
Comment (AlanFokCo) [reply vllm/device_allocator/xpumem.py]: Done — same change applied.

File: vllm/device_allocator/cumem.py (hunk 2 above)
Comment (AlanFokCo) [reply vllm/device_allocator/cumem.py]: Done — switched to `.values()` in both cumem.py and xpumem.py.

**Comment by andakai** [comment]
Thanks @AlanFokCo for the following up fix. This looks good to me now. cc @aoshen02

File: tests/basic_correctness/test_mem.py (hunk 3 above)
Comment (AlanFokCo) [reply tests/basic_correctness/test_mem.py]: Good catch — you're right, this should match sleep level 1 behavior. Changed to `offload_tags="weights"` so only weights are backed up, kv_cache is discarded as it would be in practice. The assertion still holds since weights are restored from CPU backup.

**Comment by AlanFokCo** [comment]
Thanks for the thorough review @andakai! All your suggestions made the code cleaner. 🙏

**Comment by aoshen02** [comment]
@njhill Hi nick, could you please take a quick look?

**Comment by AlanFokCo** [comment]
@njhill Hi nick, could you review this pr?

**Comment by aoshen02** [comment]
Sorry for the late response, let's try to merge it this weekend.

File: vllm/device_allocator/xpumem.py [hunk 4]
```diff
@@ -216,11 +220,38 @@ def sleep(self, offload_tags: tuple[str, ...] | str | None = None) -> None:
...
+        )
+
+        gc.collect()
+        xpu_empty_cache = getattr(torch.xpu, "empty_cache", None)
+        if callable(xpu_empty_cache):
+            xpu_empty_cache()
+
     def wake_up(self, tags: list[str] | None = None) -> None:
         for ptr, data in self.pointer_to_data.items():
             if tags is not None and data.tag not in tags:
```
Comment (aoshen02) [inline vllm/device_allocator/xpumem.py:251]: Should we also check whether weight is is_asleep to avoid wake up the tag that already up?

**Comment by andakai** [comment]
Hi @AlanFokCo , sorry for making changes directly. We want to move the kvcache release pr forward, and the review uncovered a few additional issues. So I make a new pr here https://github.com/vllm-project/vllm/pull/52514. Please let me know if you have any concerns. 

I’d also appreciate your help reviewing the follow-up kvcache release changes.

**Comment by aoshen02** [comment]
Hi Alan, we've merged it via 52514 and solved some problems and list you as first author, hope it works for you and lmk if you have any other problems.

**Comment by AlanFokCo** [comment]
Hi @aoshen02 @andakai — thanks for pushing this forward and getting it merged!

I've gone through the diff in #52514: the sleep/discard policy-conflict checks and having `wake_up()` remap only the allocations that are actually asleep are nice improvements. No concerns at all from my side — and no need to apologize for making changes directly @andakai, thanks for picking this up and carrying it over the line. 

We'll verify the merged version against our downstream use case and let you know if anything comes up. I'm also happy to help review the follow-up kvcache release changes. On our end, we plan to keep contributing in this area and will submit a few related follow-up PRs on top of this: the `_ROPE_DICT` cleanup fix, a cross-architecture sleep/wake test, and a fix for LoRA adapter handling across sleep/wake cycles.

Thanks again, everyone! 🙏

**Comment by aoshen02** [comment]
> Hi @aoshen02 @andakai — thanks for pushing this forward and getting it merged!
> 
> I've gone through the diff in #52514: the sleep/discard policy-conflict checks and having `wake_up()` remap only the allocations that are actually asleep are nice improvements. No concerns at all from my side — and no need to apologize for making changes directly @andakai, thanks for picking this up and carrying it over the line.
> 
> We'll verify the merged version against our downstream use case and let you know if anything comes up. I'm also happy to help review the follow-up kvcache release changes. On our end, we plan to keep contributing in this area and will submit a few related follow-up PRs on top of this: the `_ROPE_DICT` cleanup fix, a cross-architecture sleep/wake test, and a fix for LoRA adapter handling across sleep/wake cycles.
> 
> Thanks again, everyone! 🙏

Thanks, would be nice if you can review the kvcache release api pr, we try to get it merge earlier this week.
